File: DSA/queue.cpp
#include <bits/stdc++.h>
using namespace std;

// Creating Node Class

class Node
{
public:
    int data;
    Node *next;

    Node(int data)
    {
        this->data = data;
        next = NULL;
    }
};

// function to print meassage

void print(string message)
{
    cout << endl
         << message;
}

// Creating stack class with its functions

class Queue
{
public:
    Node *front;
    Node *rear;

    Queue()
    {
        front = NULL;
        rear = NULL;
    }

    // for enqueue operation

    void enqueue()
    {

        // Taking input from the user
        int data;
        print("Enter the data: ");
        cin >> data;

        Node *newNode = new Node(data);

        if (rear == NULL)
        {
            rear = front = newNode;
            return;
        }

        rear->next = newNode;
        rear = newNode;
    }

    // Function for dequeue Operation

    void dequeue()
    {
        if (front == NULL)
        {
            return;
        }

        Node *temp = front;
        front = front->next;

        if (front == NULL)
        {
            rear = NULL;
        }

        delete temp;
    }

    // Update Function

    void updateWithKey()
    {
        // Taking input from the user

        int key;
        print("Enter the key: ");
        cin >> key;
        int data;
        print("Enter the data: ");
        cin >> data;

        Node *temp = front;

        while (temp->data != key && temp != NULL)
        {
            temp = temp->next;
        }

        if (temp == NULL)
        {
            print("No such key present.");
        }
        else
        {
            temp->data = data;
        }
    }

    // Search Function

    void search()
    {

        int data;
        print("Enter the data to search: ");
        cin >> data;
        Node *temp = front;

        while (temp != NULL)
        {
            if (temp->data == data)
            {
                cout << endl
                     << "Present";
                return;
            }
            temp = temp->next;
        }

        print("Not Present");
    }

    // to display

    void display()
    {
        Node *temp = front;

        if (!temp)
        {
            print("Queue is empty");
        }
        cout << endl;
        while (temp != NULL)
        {
            cout << temp->data;
            temp = temp->next;
            if (temp)
            {
                cout << "->";
            }
        }
    }
};

int main()
{
    // Creating queue Object

    Queue q;

    // while for menu --

    while (true)
    {
        print("Your Queue -- ");
        q.display();

        cout << endl;

        print("Enter 1 for enqueue operation");
        print("Enter 2 for dequeue operation");
        print("Enter 3 for updateWithKey operation");
        print("Enter 4 for search operation");

        print("Enter 5 to exit");
        cout << endl;
        int input;
        cin >> input;
        int data;

        switch (input)
        {
        case 1:
            q.enqueue();
            break;
        case 2:
            q.dequeue();
            break;
        case 3:
            q.updateWithKey();
            break;
        case 4:
            q.search();
            break;

        default:
            return 0;
        }
    }
}

File: Data Structures/Queue.java
/*
 * Implementation of a linked list for Integers.
 *  This can be also be converted into a generic LinkedList by changing the
 *  type of the data to a generic type.
 */


//Importing Scanner class for user input
import java.util.Scanner;

public class Queue {
    static class Node {
        //        Data of the node
        int data;
        // Pointer to the next node
        Node next;
        Node(int data){
            this.data = data;
            this.next = null;
        }
    }
    //    Create a scanner to take input from the user
    static Scanner sc = new Scanner(System.in);
    public static void main(String[] args) {
        // Create a blank linked list
        Node head = null;
        // Variable to store the choice of the user
        int choice;
        do {
            //  Printing the menu for the user
            System.out.println(" --------- MENU --------- ");
            System.out.println("Press 0 : Create a New Queue");
            System.out.println("Press 1 : Enqueue an element");
            System.out.println("Press 2 : Dequeue an element");
            System.out.println("Press 3 : Update an element");
            System.out.println("Press 4 : Display/Traversal the LinkedList");
            System.out.println("Press 5 : Exit");
            System.out.println("Enter your choice : ");
            choice = sc.nextInt();
            //  Taking the choice from the user and performing the corresponding operation
            switch (choice) {
                case 0:
                    System.out.print("Please Enter the first Node Value: ");
                    head = createQueue(sc.nextInt());
                    break;
                case 1:
                    if (head != null) {
                        System.out.print("Please Enter the Node Value: ");
                        head = enqueue(head, sc.nextInt());
                    } else {
                        System.out.println("Please create a LinkedList first");
                    }
                    break;
                case 2:
                    System.out.println("Please Enter the Node Value: ");
                    int dequeuedElement = dequeue(head);
                    break;
                case 3:
                    System.out.print("Please Enter the Node Value: ");
                    updateElement(head, sc.nextInt());
                    break;
                case 4:
                    display(head);
                    break;
                case 5:
                    System.out.println("Program Terminated Successfully");
                    break;
                default:
                    System.out.println("Invalid Input");
            }
        } while (choice != 5);

    }


    private static Node enqueue(Node head, int nextInt) {
        Node temp = head;
        while (temp.next != null) {
            temp = temp.next;
        }
        temp.next = new Node(nextInt);
        return head;
    }
    private static int dequeue(Node head) {
        if(head == null){
            System.out.println("Queue is empty");
            return -1;
        }
        int dequeuedElement = head.data;
        head = head.next;
        return dequeuedElement;
    }
    private static void updateElement(Node head, int nextInt) {
        Node temp = head;
        while (temp != null) {
            if (temp.data == nextInt) {
                System.out.print("Please Enter the new Node Value: ");
                temp.data = sc.nextInt();
                break;
            }
            temp = temp.next;
        }
    }

    private static void display(Node head) {
        Node temp = head;
        System.out.println("The Queue is: ");
        while (temp != null) {
            System.out.print(temp.data + " ");
            temp = temp.next;
        }
        System.out.println();
    }

    private static Node dequeue(Node head, int nextInt) {
        if(head ==null){
            System.out.println("Queue is empty");
            return null;
        }
        else{
            Node temp = head;
            head = head.next;
            System.out.println("The deleted element is: " + temp.data);
            return head;
        }
    }

    private static Node createQueue(int nextInt) {
      Node head = new Node(nextInt);
        return head;
    }


}


